common/web/gesture-recognizer/src/pathSegmenter.ts

```diff
@@ -0,0 +1,955 @@
+        // The higher this is, the more we like its potential as the segmentation point.
+        get segRating() {
+          if(this.candidate) {
+            return Math.max(this.xTest.coefficientOfDetermination, this.yTest.coefficientOfDetermination);
+          } else {
+            return 0;
+          }
+        }
+
+        // But, if it turns out this potential point wouldn't actually result in segmentation, well...
+        // "abandon ship".
+        get segmentationMerited() {
+          return this.candidate?.segmentationMerited ?? false;
+        }
+      }
+
+      // Start:  split-point search
+
+      /* We'll use our initial sliding window as a starting point.  It's close to the active location
+       * of the touch, so we're unlikely to miss a proper segmentation point if it lies close.
+       * We'll find the local maximum (best split point for the local part of the path)
+       * via gradient descent.
+       *
+       * Note:  binary search is a 'bad idea', as the touchpath is likely not following a
+       * monotonous path, mathematically speaking.  (It'd be possible to overshoot a "peak" or
+       * "valley" quite easily.)
+       */
+      let currentSplit = new SplitSearchState(candidateSplit);
+
+      let leftSplit = new SplitSearchState(new PotentialSegmentation(this.steppedCumulativeStats, this.choppedStats, splitPoint-1));
+      let rightCandidate: PotentialSegmentation = null;
+      if(splitPoint+1 < this.steppedCumulativeStats.length) {
+        rightCandidate = new PotentialSegmentation(this.steppedCumulativeStats, this.choppedStats, splitPoint+1);
+      }
+      let rightSplit = new SplitSearchState(rightCandidate);
+
+      // Step 2a:  detect which direction gives the best improvement in segmentation potential.
+      const criteria = [leftSplit.segRating, currentSplit.segRating, rightSplit.segRating];
+      let sortedCriteria = [...criteria].sort();
+
+      const delta = criteria.indexOf(sortedCriteria[2])-1;  // -1 if 'left' is best, 1 if 'right' is best.
+
+      // Step 2b: we've found the direction:  go searching!
+      if(delta != 0) {
+        // We can get better segmentation by shifting.  Proceed in the optimal direction.
+        do {
+          const nextSplitIndex = splitPoint + delta;
+          if(nextSplitIndex >= this.steppedCumulativeStats.length || nextSplitIndex < 0) {
+            break;
+          }
+
+          let nextCandidate = new PotentialSegmentation(this.steppedCumulativeStats, this.choppedStats, splitPoint + delta);
+          let nextSplit = new SplitSearchState(nextCandidate);
+
+          // Prevent overly-short intervals / over-segmentation.
+          if(nextCandidate.pre.duration * 1000 < this.SLIDING_WINDOW_INTERVAL / 2) {
+            break;
+          } else if(nextCandidate.post.duration * 1000 < this.SLIDING_WINDOW_INTERVAL / 2) {
+            break;
+          }
```

Comment:
- [ ] Just noticed that I neglected to fix this section when I shifted `.duration` from seconds to milliseconds.